src/language-html/printer-html.js

```diff
@@ -377,112 +378,206 @@ function genericPrint(path, options, print) {
-            leadingParts.push(prevBetweenLine);
+        const nextBetweenLine = childNode.next
+          ? printBetweenLine(childNode, childNode.next)
+          : "";
+
+        if (prevBetweenLine) {
+          if (forceNextEmptyLine(childNode.prev)) {
+            prevParts.push(hardline, hardline);
+          } else if (prevBetweenLine === hardline) {
+            prevParts.push(hardline);
           } else {
-            leadingParts.push(
-              ifBreak("", softline, {
-                groupId: groupIds[childIndex - 1]
-              })
-            );
+            if (isTextLikeNode(childNode.prev)) {
+              leadingParts.push(prevBetweenLine);
+            } else {
+              leadingParts.push(
+                ifBreak("", softline, {
+                  groupId: groupIds[childIndex - 1]
+                })
+              );
+            }
           }
         }
-      }
 
-      if (nextBetweenLine) {
-        if (forceNextEmptyLine(childNode)) {
-          if (isTextLikeNode(childNode.next)) {
-            nextParts.push(hardline, hardline);
-          }
-        } else if (nextBetweenLine === hardline) {
-          if (isTextLikeNode(childNode.next)) {
-            nextParts.push(hardline);
+        if (nextBetweenLine) {
+          if (forceNextEmptyLine(childNode)) {
+            if (isTextLikeNode(childNode.next)) {
+              nextParts.push(hardline, hardline);
+            }
+          } else if (nextBetweenLine === hardline) {
+            if (isTextLikeNode(childNode.next)) {
+              nextParts.push(hardline);
+            }
+          } else {
+            trailingParts.push(nextBetweenLine);
           }
-        } else {
-          trailingParts.push(nextBetweenLine);
         }
-      }
 
-      return concat(
-        [].concat(
-          prevParts,
-          group(
-            concat([
-              concat(leadingParts),
-              group(concat([printChild(childPath), concat(trailingParts)]), {
-                id: groupIds[childIndex]
-              })
-            ])
-          ),
-          nextParts
-        )
-      );
-    }, "children")
+        return concat(
+          [].concat(
+            prevParts,
+            group(
+              concat([
+                concat(leadingParts),
+                group(concat([printChild(childPath), concat(trailingParts)]), {
+                  id: groupIds[childIndex]
+                })
+              ])
+            ),
+            nextParts
+          )
+        );
+      }, "children")
+      .filter(res => res !== "ignore")
   );
 
+  /**
+   * check the ignoreRanges for the given index.
+   * this will either return a Doc to be included in the output, 'ignore' if
+   * the given child should *not* be included in the output, or 'continue' if
+   * the child should process normally.
+   * @return {'ignore' | 'continue' | Doc}
+   */
+  function maybeIgnore(path, index, ignoreRanges) {
```

Comment:
Let's move this to `util`

Reply:
It's a bit difficult to move this as it uses variables from outer scopes and the doc builders. Other functions in utils don't use the doc builders. Let's keep it as is.